gst/mqtt/mqttsrc.c

```diff
@@ -61,6 +62,7 @@ enum
   DEFAULT_MQTT_OPT_KEEP_ALIVE_INTERVAL = 60,    /* 1 minute */
   DEFAULT_MQTT_SUB_TIMEOUT = 10000000,  /* 10 seconds */
   DEFAULT_MQTT_SUB_TIMEOUT_MIN = 1000000,       /* 1 seconds */
+  DEFAULT_MQTT_QOS = 2,         /* Once and one only */
```

Comment:
The reason to set this as 2? (mqtt-sink had it 0, right?)

Reply:
Yes, sink & src(pub & sub) a different default value.
According to the official document, they recommend the QoS level of subscribers as the maximum.

> ... The MQTTAsync_subscribe() and MQTTAsync_subscribeMany() functions set this maximum...
https://www.eclipse.org/paho/files/mqttdoc/MQTTAsync/html/qos.html

Reply:
Q1: What happens if QoS level of pub and sub are different? What's the behavior in the case of (pub 0 and sub2)?
Q2: What are the proper default values for our expected AI applications?

Reply:
@myungjoo Please refer http://www.steves-internet-guide.com/wp-content/uploads/MQTT-qos-retain-clean-session-table.jpg ([ref](http://www.steves-internet-guide.com/mqtt-retained-messages-example/)). It looks confusing to me. IMO, by default, QoS should be 0 (which is fire-and-forget). When I did use QoS1 or QoS2, the broker tries to keep the messages so that it is terminated by OOM in some cases or the broker repeatedly the same message to all subscribers if one of them fails to get the message from the same topic. More investigation is needed.

Reply:
@wooksong Then, do you think default-QoS should be 0 for sub as well?

Reply:
@myungjoo I do. However, when I did some tests about it, there are many variables that might influence the results. We need to investigate it using the recent MQTT elements.

Reply:
In my understanding, the qos-level of each sinks&srcs are working respectively only with broker.

>The QoS Level is restricted to client to broker or broker to client communication. So if I publish a message with QoS Level 2, the message guarantee is only applicable from THAT client to the broker. If another client is subscribed to a topic on that broker with QoS Level 0, then the broker will send messages to that client only with QoS Level 0.
>> https://mntolia.com/mqtt-qos-levels-explained/

Reply:
The QoS level of the whole network follows the minimum level.
Furthermore, if all the QoS of all subscribers are 2(the highest), what users have to do to change the QoS of the whole network is to change the level of the publisher(mqttsink). 
As I know, this is the reason why they recommend 2(the highest) as a default QoS level of subscribers.

Reply:
>The QoS of a message forwarded to a subscriber thus might be different to the QoS given to the message by the original publisher. The lower of the two values is used to forward a message.
>>https://www.eclipse.org/paho/files/mqttdoc/MQTTClient/html/qos.html

Reply:
I attached the log of `mosquitto` broker for all cases of QoS level of 1 sub & 1pub
According to the below result, the QoS level of the publisher is the standard.

[pub QoS - sub QoS]
<details>
<summary>[0-0], [0-1], [0-2] (same)</summary>
<pre>
1620990738: Sending CONNACK to mosqpub|2970-camsrc1 (0, 0)
1620990738: Received PUBLISH from mosqpub|2970-camsrc1 (d0, q0, r0, m0, 'qostest', ... (1 bytes))
1620990738: Sending PUBLISH to mosqsub|2923-camsrc2 (d0, q0, r0, m0, 'qostest', ... (1 bytes))
1620990738: Received DISCONNECT from mosqpub|2970-camsrc1
1620990738: Client mosqpub|2970-camsrc1 disconnected.
</pre>
</details>

<details>
<summary>[1-0]</summary>
<pre>
1620990790: Sending CONNACK to mosqpub|3064-camsrc1 (0, 0)
1620990790: Received PUBLISH from mosqpub|3064-camsrc1 (d0, q1, r0, m1, 'qostest', ... (1 bytes))
1620990790: Sending PUBACK to mosqpub|3064-camsrc1 (Mid: 1)
1620990790: Sending PUBLISH to mosqsub|3013-camsrc2 (d0, q0, r0, m0, 'qostest', ... (1 bytes))
1620990790: Received DISCONNECT from mosqpub|3064-camsrc1
1620990790: Client mosqpub|3064-camsrc1 disconnected.
</pre>
</details>

<details>
<summary>[1-1] [1-2] (same)</summary>
<pre>
1620990807: Sending CONNACK to mosqpub|3066-camsrc1 (0, 0)
1620990807: Received PUBLISH from mosqpub|3066-camsrc1 (d0, q1, r0, m1, 'qostest', ... (1 bytes))
1620990807: Sending PUBACK to mosqpub|3066-camsrc1 (Mid: 1)
1620990807: Sending PUBLISH to mosqsub|3014-camsrc2 (d0, q1, r0, m1, 'qostest', ... (1 bytes))
1620990807: Received DISCONNECT from mosqpub|3066-camsrc1
1620990807: Client mosqpub|3066-camsrc1 disconnected.
1620990807: Received PUBACK from mosqsub|3014-camsrc2 (Mid: 1)
</pre>
</details>

<details>
<summary>[2-0]</summary>
<pre>
1620990691: Sending CONNACK to mosqpub|2900-camsrc1 (0, 0)
1620990691: Received PUBLISH from mosqpub|2900-camsrc1 (d0, q2, r0, m1, 'qostest', ... (1 bytes))
1620990691: Sending PUBREC to mosqpub|2900-camsrc1 (Mid: 1)
1620990691: Received PUBREL from mosqpub|2900-camsrc1 (Mid: 1)
1620990691: Sending PUBCOMP to mosqpub|2900-camsrc1 (Mid: 1)
1620990691: Sending PUBLISH to mosqsub|2836-camsrc2 (d0, q0, r0, m0, 'qostest', ... (1 bytes))
1620990691: Received DISCONNECT from mosqpub|2900-camsrc1
1620990691: Client mosqpub|2900-camsrc1 disconnected.
</pre>
</details>

<details>
<summary>[2-1]</summary>
<pre>
1620990664: Sending CONNACK to mosqpub|2878-camsrc1 (0, 0)
1620990664: Received PUBLISH from mosqpub|2878-camsrc1 (d0, q2, r0, m1, 'qostest', ... (1 bytes))
1620990664: Sending PUBREC to mosqpub|2878-camsrc1 (Mid: 1)
1620990664: Received PUBREL from mosqpub|2878-camsrc1 (Mid: 1)
1620990664: Sending PUBCOMP to mosqpub|2878-camsrc1 (Mid: 1)
1620990664: Sending PUBLISH to mosqsub|2834-camsrc2 (d0, q1, r0, m1, 'qostest', ... (1 bytes))
1620990664: Received DISCONNECT from mosqpub|2878-camsrc1
1620990664: Client mosqpub|2878-camsrc1 disconnected.
1620990664: Received PUBACK from mosqsub|2834-camsrc2 (Mid: 1)
</pre>
</details>

<details>
<summary>[2-2]</summary>
<pre>
1620990628: Sending CONNACK to mosqpub|2792-camsrc1 (0, 0)
1620990628: Received PUBLISH from mosqpub|2792-camsrc1 (d0, q2, r0, m1, 'qostest', ... (1 bytes))
1620990628: Sending PUBREC to mosqpub|2792-camsrc1 (Mid: 1)
1620990628: Received PUBREL from mosqpub|2792-camsrc1 (Mid: 1)
1620990628: Sending PUBCOMP to mosqpub|2792-camsrc1 (Mid: 1)
1620990628: Sending PUBLISH to mosqsub|2749-camsrc2 (d0, q2, r0, m1, 'qostest', ... (1 bytes))
1620990628: Received DISCONNECT from mosqpub|2792-camsrc1
1620990628: Client mosqpub|2792-camsrc1 disconnected.
1620990628: Received PUBREC from mosqsub|2749-camsrc2 (Mid: 1)
1620990628: Sending PUBREL to mosqsub|2749-camsrc2 (Mid: 1)
1620990628: Received PUBCOMP from mosqsub|2749-camsrc2 (Mid: 1)
</pre>
</details>

Reply:
@helloahn and I discussed the default values of QoS parameters. As it is officially recommended, the default values of the QoS parameters are 0 and 2 for the publisher and subscriber, respectively.

Reply:
the details have been attached at https://github.com/nnstreamer/nnstreamer/pull/3271#issuecomment-842096639.